product_variant_exclusion/models/product.py

```diff
@@ -0,0 +1,174 @@
+                        raise UserError(_(
+                            'You cannot leave only one attribute value on a '
+                            'line if this value is part of an exclusion'))
+        if self.value_ids.ids != attr_values:
+            new_values = self.env[
+                'product.attribute.value'].browse(attr_values)
+            exclusion_values = self.product_tmpl_id.variant_exclusion_ids\
+                .mapped('attribute_value_ids')
+            for new_val in new_values:
+                if new_val in exclusion_values:
+                    raise UserError(_('You cannot remove an attribute value '
+                                      'from an attribute line, if this value '
+                                      'is part of an exclusion.'))
+        return super(ProductAttributeLine, self).write(vals)
+
+    @api.multi
+    def unlink(self):
+        for line in self:
+            for val_id in line.value_ids.ids:
+                if val_id in line.product_tmpl_id.variant_exclusion_ids.mapped(
+                        'attribute_value_ids').ids:
+                    raise UserError(_(
+                        'You cannot delete an attribute line when '
+                        'one of its values is part of an exclusion'))
+        return super(ProductAttributeLine, self).unlink()
+
+
+class ProductVariantExclusion(models.Model):
+
+    _name = 'product.variant.exclusion'
+
+    product_tmpl_id = fields.Many2one('product.template', required=True)
+    attribute_value_ids = fields.Many2many('product.attribute.value',
+                                           required=True)
+
+    def _check_values_attributes_uniqueness(self, value_ids):
+        values = self.env['product.attribute.value'].browse(value_ids)
+        attributes = values.mapped('attribute_id')
+        if len(values) != len(attributes):
+            raise UserError(_('You cannot have multiple values from the '
+                              'same attribute in an exclusion.'))
+
+    @api.model
+    def create(self, vals):
+        value_ids = vals.get('attribute_value_ids')[0][2]
+        if len(value_ids) <= 1:
+            raise UserError(_(
+                'You cannot create an exclusion with only one value !'))
+        template_lines = self.env['product.attribute.line'].search([
+            ('product_tmpl_id', '=', vals.get('product_tmpl_id')),
+            ('value_ids', 'in', value_ids)])
+        for line in template_lines:
+            if len(line.value_ids) <= 1 or set(line.value_ids.ids) <= set(
+                    value_ids):
+                raise UserError(_(
+                    'You cannot create an exclusion for an attribute value '
+                    'which is unique in its attribute line !'))
+        self._check_values_attributes_uniqueness(value_ids)
+        return super(ProductVariantExclusion, self).create(vals)
+
+    @api.multi
+    def write(self, vals):
+        value_ids = vals.get('attribute_value_ids')[0][2]
```

Comment:
Also I wonder if a lot of these write and create overrides aren't better dealt with @api.constrains().  Maybe a technical person can enlighten it more, but these seem like constraints rather than changing anything.